flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java

```diff
@@ -861,6 +869,43 @@ else if (transitionState(current, ExecutionState.FAILED, t)) {
 		}
 	}
 
+	private void waitInvokableCancelCompletion() throws java.util.concurrent.ExecutionException, java.util.concurrent.TimeoutException {
+		if (!invokableHasBeenCanceled.get()) {
+			return;
+		}
+
+		// set deadline if taskCancellationTimeout is set to have time for cleanup (before being killing by TaskCancelerWatchDog) (best effort)
+		Optional<Deadline> deadline = Optional
+			.of((long) (taskCancellationTimeout * .9))
+			.filter(t -> t > 0)
+			.map(t -> Deadline.fromNow(Duration.ofMillis((t))));
```

Comment:
What's the motivation behind this deadline? I think it should be perfectly fine/safe to simplify this and just wait on the future indefinitely, as this is the behaviour for network tasks and for source tasks before we introduced the mailbox (in the past, `SourceFunction#run()` loop was called from the task thread).

Reply:
The motivation was to preserve the *current* behaviour and have time for cleanup which includes:
```
releaseResources();
memoryManager.releaseAll(invokable);
fileCache.releaseJob(jobId, executionId);
FileSystemSafetyNet.closeSafetyNetAndGuardedResourcesForThread();
notifyFinalState();
```

However, I see your point (especially given that old behavior didn't have a thread leak issue) so I'll remove the deadline.